code/network/multi_turret_manager.h

```diff
@@ -0,0 +1,96 @@
+    
+    turret_packet(int time_in, ushort target_netsig, std::pair<bool, float> ang1_in, std::pair<bool, float> ang2_in) 
+    {
+        time = _timestamp(time_in + Multi_Timing_Info.get_mission_start_time());
+        target_objnum = OBJ_INDEX(multi_get_network_object(target_netsig));
+    
+        angle1.first = ang1_in.first;
+
+        if (ang1_in.first){
+            angle1.second = ang1_in.second;
+        } else {
+            angle1.second = 0.0f;
+        }
+
+        angle2.first = ang2_in.first;
+
+        if (ang2_in.first){
+            angle2.second = ang2_in.second;
+        } else {
+            angle2.second = 0.0f;
+        }
+    }
+};
+
+// basic hash function to allow us to use turret_packet_address as a key for unordered map
+struct turret_packet_address_hash // <table_entry_option_address>
+{
+	std::size_t operator()(const turret_packet_address& k) const noexcept
+	{
+		// Compute individual hash values and combine them using XOR and bit shifting:
+		return ( (std::hash<int>{}(static_cast<int>(k.first)))
+				^ (std::hash<int>{}(k.second) << 1));
+	}
+};
+
+class multiplayer_turret_manager {
+    SCP_unordered_map<turret_packet_address, SCP_vector<turret_packet>, turret_packet_address_hash> _packets_in;
+    SCP_unordered_map<ushort, SCP_vector<int>> _packets_out;
+
+public:
+    void add_incoming_packet(int time_in, int parent_ship, short subsys_index, ushort target_netsig, std::pair<bool, float> ang1_in, std::pair<bool, float> ang2_in) 
+    {
+        // if the data is nonsense, return
+        if (parent_ship < 0 || parent_ship > MAX_OBJECTS || Objects[parent_ship].net_signature == 0){
+            return;
+        }
+
+        _packets_in[turret_packet_address(Objects[parent_ship].net_signature, static_cast<int>(subsys_index))].emplace_back(time_in, target_netsig, ang1_in, ang2_in);
+    }
+
+    bool get_latest_packet(ushort net_signature, int subsys_index, turret_packet* packet);
+
+    void reset();
+
+    void add_packet_to_queue(ushort net_signature, int subsys_index) {
+
+        // if the subsysem actually can't be sent, don't queue it.
+        if (net_signature == 0  || subsys_index > INT16_MAX) {
+            nprintf(("network", "Trying to queue a turret tracking packet with unsendable info: net_signature %d and subsys_index %d.", net_signature, subsys_index));
+            return;
+        }
+        
+        // if this one was already added, there's no need to add it again.  
+        if (std::find(_packets_out[net_signature].begin(), _packets_out[net_signature].end(), subsys_index) != _packets_out[net_signature].end()){
+            return;
+        }
+
+        _packets_out[net_signature].push_back(subsys_index);
+    }
+
+    void send_queued_packets(){
+        for (auto& queued : _packets_out) {
+            extern SCP_vector<int> send_turret_tracking_packet(ushort parent_sig, SCP_vector<int>& subsys_indexes);
+            queued.second = send_turret_tracking_packet(queued.first, queued.second);
```

Comment:
You should verify that `queued.second` actually has data before processing the packet. As `_packets_out` still contains all of the packets which have previously been processed, it's loaded with useless data which is just wasting time and bandwidth.

Reply:
Yes, you're right that I should check if it's empty so that we don't end up sending an empty packet.

But part of the design of `send_turret_tracking_packet` is to return only the subsystems that there wasn't room for in the packet for the manager to reattempt later.  So it won't contain the packets already sent.  It will still have the ship in the unordered map, but the vector in most cases will be empty after the packet is sent.